Move a lot of variable data into tag storage

lindsayad opened the pull request:
And the required coupleable interfaces

Closes #20586

lindsayad committed:
Create TagVectoryArrayAux

And the required coupleable interfaces

Closes #20586

lindsayad commented:
@fdkong I would like to be able to change the `vector_tag` in this test to `residual` but I can't because it is not associated during `timestep_end`. It would be cool if we could enable that capability somehow. What do you think?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1024261/) on 3f3c3fd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20841/site)

This comment will be updated on new commits.

fdkong commented:
> I would like to be able to change the vector_tag in this test to residual

Which test? I did not get it yet

fdkong commented:
OK, I see what you meant now. The issue is that the `residual` vector is owned by PETSc solvers. We update it in the computeResidual function. The residual vector was passed from PETSc side as an input parameter of computeResidual. Once we are done with computeResidual,  we need to safely return the residual back to PETSc without any further modification. From the software design point of view, it might not be a great idea to access that vector.

What are use cases here? What we could do is that we can have a MOOSE "save-in" residual vector that will have the same values as the residual vector, and we access that vector whenever we want. The downside is increase memory usage

lindsayad commented:
What about for nonlinear systems, we can access the [ExplicitSystem rhs member](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1ExplicitSystem.html#a98fde3a02d1256dbc31891c88986bddc)? I believe that should correspond to the residual that PETSc sends us right? Because we call `SNESSetFunction` with it in `PetscNonlinearSolver::solve` (through `NonlinearImplicitSystem::solve`)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1026436/) on f22eac9 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a34443</th>
      <th colspan=4>#20841 f22eac</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.61%</td>
      <td>82.62%</td>
      <td>+0.00%</td>
      <td>77.27%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>74483</td>
      <td>74507</td>
      <td>+24</td>
      <td>51</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15677</td>
      <td>15677</td>
      <td>-</td>
      <td>15</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20841/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20841/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20841/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 77.27% is less than the suggested 90.0%



This comment will be updated on new commits.

fdkong commented:
Yeah, for that case, it might be correct. But that might not work in general, Eigensolver might not follow that.

lindsayad commented:
Well, I don't think I need it right this instant. But in the future it can be very nice to visualize the whole residual vector in an easy way. I know that we can do that by adding an `extra_tag_vectors` and then adding `extra_vector_tags` to all residual objects, so maybe I am fine :smile:

YaqiWang commented on framework/src/auxkernels/TagVectorArrayAux.C:
```diff
@@ -0,0 +1,47 @@
+}
+
+Real
+TagVectorArrayAux::computeValue()
```
Should override `compute()` like `BuildArrayVariableAux` to directly manipulate dofs.

lindsayad replied:
I should not have to do that, right? I should be able to still just override `computeValue` like `FunctionArrayAux` does, right?

YaqiWang replied:
It is more costly. Implementing `compute()` is quite simple.

lindsayad replied:
It is strange when you bring up cost. When I am reviewing your code, you are not worried about cost, and then when you are reviewing mine, you are worried about cost. :laughing: Have you profiled this? Are the 9x more lines giving you much speed up? It is sure nice to just return `_v[_qp]` in a single-line implementation

YaqiWang replied:
I typically care simpler codes but I care more whether we are doing things in a more direct way, i.e. if data are held in the right places, if we activates unnecessary code pathes, etc. Here manipulating dofs is the direct way. A by-product is the performance because it by pass the local mass matrix inversion (I do not care this much).

lindsayad replied:
I just get really tired of your seeming double standards. You always demand that we review your code quickly and you always push back when we suggest changes. "Unnecessary code paths". You mean like writing a one-to-all algorithm in a few lines of code as opposed to an all-to-all in a single line?

lindsayad replied:
I will make your suggested change

lindsayad replied:
Actually it seems to me like your suggested change will not work unless the array variable is Lagrange. We would need to make a tag version of `coupledAllDofValues` in order for your change to make sense

YaqiWang replied:
Yeah, we always push back. I understand. I have leant to listen reviewers ;-)

YaqiWang replied:
Wait, I am checking.

YaqiWang replied:
That is true, we only have `coupledArrayDofValues` but not `coupledVectorTagArrayDofValues` yet. Maybe add that? Sorry for this.

lindsayad replied:
I will probably delete `coupledVectorTagArrayValue` then, since there will still be no use for it

YaqiWang commented on framework/src/auxkernels/TagVectorArrayAux.C:
```diff
@@ -0,0 +1,47 @@
+}
+
+TagVectorArrayAux::TagVectorArrayAux(const InputParameters & parameters)
+  : AuxKernel(parameters),
```
Why not make this as an ArrayAuxKernel?

lindsayad replied:
Do we automatically handle multiple components in our libMesh `ExodusIO`? If so, then I will change this to not take a `component` parameter and make it an ArrayAuxKernel

YaqiWang replied:
Yes, because a component of an array variable is a normal libMesh variable. Just all components are in one variable group.

YaqiWang commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -251,6 +251,19 @@ class Coupleable
+   * @return Reference to a VariableValue for the coupled variable
+   * @see Kernel::_u
+   */
+  const ArrayVariableValue & coupledVectorTagArrayValue(const std::string & var_name,
```
@fdkong this can be on any tagged vector, not necessarily on a residual vector.

YaqiWang reviewed (changes requested):
I think @zachmprince wants to have this for using save-in to evaluate a postprocessor required by Griffin.

lindsayad committed:
Use ArrayAuxKernel instead of AuxKernel for TagVectorArrayAux

lindsayad committed:
Test coupledVectorTagValues API

lindsayad committed:
Implement Coupleable::vectorTagDofValueHelper

In order to make this work I had to make a virtual interface at the
MooseVariableField level for vectorTagDofValue, which led to me
implementing that method for finite volume variables, and then testing
it. In testing, I discovered that it can be dangerous to add old
solution state tags to the Coupleable _fe_coupleable_vector_tags
member. The danger comes from the fact that members like _vector_tag_u
in MooseVariableData(FV) are sized at construction time and this may
happen before tags for old solution states are added. So if we attempt
to index into _vector_tag_u with the old solution state tag, this can
lead to out-of-bounds access. However, not to fear... there is no
conceptual need for having an entry in _vector_tag_u for old solution
states for which we already have dedicated data members like _dof_values_old,
_dof_values_older, _dof_values_previous_nl. So I created a _special_tags
data member in Coupleable corresponding to these data types which
allows us to avoid indexing _vector_tag_u out-of-bounds

Other notes:
- Made solution state methods in SystemBase virtual such that the
  displaced systems can call their undisplaced system counterpart
  methods
- Fixed a bug in vectorTagValue in the variable classes where we were
  checking for OLD in the OLDER branch

YaqiWang commented:
How do I approve this?

lindsayad commented:
You should be able to review changes and approve. It'll give a gray check
mark. We'll need someone else's approval too but your approval should be a
good guide

On Thu, Apr 28, 2022 at 6:49 PM Yaqi ***@***.***> wrote:

> How do I approve this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/20841#issuecomment-1112808273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4GB6OMR6LPCACEIFCDVHM5ZVANCNFSM5UAPG3VA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

lindsayad commented:
@fdkong do you want to give this a look?

YaqiWang commented:
I possibly do not have right to approve PRs in MOOSE. I do not see a gray check mark but this:
![Screen Shot 2022-05-02 at 8 49 12 AM](https://user-images.githubusercontent.com/6607496/166254930-7cdf8be3-9569-4b18-ac5d-6ae0f881cdfe.png)
Maybe I was missing something?

fdkong commented on framework/doc/content/source/auxkernels/TagVectorArrayAux.md:
```diff
@@ -0,0 +1,17 @@
+auxiliary variable that this object populates must match the family and order of
+the array variable specified through `v`. Note that the only allowed execute-on
+option for this object is `timestep_end` since residual evaluation must be
+completed before reliable indexing of the tagged vector can be performed.
```
We can relax this if the tagged vector is a solution vector.

lindsayad replied:
do we have an api to check whether a tag is a solution or residual tag?

fdkong commented on framework/src/variables/MooseVariableDataFV.C:
```diff
@@ -1143,6 +1143,55 @@ MooseVariableDataFV<OutputType>::dofValuesDuDotDotDu() const
+    return sln(Moose::Old);
+  else if (_subproblem.vectorTagExists(Moose::OLDER_SOLUTION_TAG) &&
+           tag == _subproblem.getVectorTagID(Moose::OLDER_SOLUTION_TAG))
+    return sln(Moose::Older);
```
It would be helpful if you could explain why we need to special handles here for `OLDER_SOLUTION_TAG` and others. I feel like we want to handle them in an uniform way if that is possible

lindsayad replied:
I am just mimicking what we already have in `MooseVariableData`. But there is a problem with the tagging system linked to this code. We size `_vector_tag_u` and related data members in the variable constructors based off the current number of vector tags in the `_subproblem` at that time. However, older solution states may not be requested until later. This can cause indexing out-of-bounds in `_vector_tag_u` unless you use things like `sln(Moose::OLD)` and `sln(Moose::OLDER)`. We could change 
```c++
  else
  {
    _need_vector_tag_u[tag] = true;

    if (_sys.hasVector(tag) && tag < _vector_tag_u.size())
      return _vector_tag_u[tag];
    else
      mooseError("Tag is not associated with any vector or there is not any data for tag ",
                 tag,
                 " for variable ",
                 _var.name());
  }
```
to something like
```c++
  else
  {
    if (tag >= _vector_tag_u.size())
      {
        _need_vector_tag_u.resize(tag + 1, false);
        _vector_tag_u.resize(tag + 1);
      }
      
    _need_vector_tag_u[tag] = true;

    if (_sys.hasVector(tag))
      return _vector_tag_u[tag];
    else
      mooseError("Tag ", tag, " is not associated with any vector for variable ",
                 _var.name());
  }
```
and then we could just `_vector_tag_u` for everything and not do special dispatch for certain solution vector tags. What do you think?

fdkong replied:
I like this idea better.

fdkong replied:
This will make code clearer and simpler

lindsayad replied:
this has not been simpler ... this has turned into a monstrous task :frowning_face:

lindsayad committed:
Encompass sln/dof/grad variable data members in tag storage

This was a pretty big change and I took this "opportunity" to create
a `MooseVariableDataBase` class such that I wasn't refactoring
two different classes (FE and FV) in the exact same way

lindsayad commented:
@fdkong this should be ready for further review. I already have a fix for the Magpie failure at idaholab/magpie#476

moosebuild commented:
Job [External app tests](https://civet.inl.gov/job/1039617/) on e5a1f6b : invalidated by @lindsayad

fdkong commented on framework/include/auxkernels/TagVectorArrayAux.h:
```diff
@@ -0,0 +1,36 @@
+ * Couple a tagged vector, and return its evaluations at degree of freedom indices corresponding to
+ * the coupled array variable
+ */
+class TagVectorArrayAux : public TagAuxBase<ArrayAuxKernel>
```
Similar for all other places if we decide to make that change

fdkong reviewed (changes requested):
Did not get the review done yet. Just a few comments at this point 

I need to dig into the changes more later. Just let you know I am working on this

lindsayad commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -1458,6 +1505,10 @@ class Coupleable
   const MooseObject * const _obj;
+
+  /// vector tag names for which we need to request older solution states from the system
+  const std::set<std::string> _older_state_tags = {Moose::OLD_SOLUTION_TAG,
```
I added these in order to be able to recognize tags for which we would need to request additional solution states from the `SystemBase`. Otherwise a user who asks for an old/older tag may get an error that the vector tag does not exist in the subproblem/system

fdkong replied:
Did you mean a user explicitly ask for an old/older tag? I would imagine these tags will be required by transient executioners or/and time integrators.

lindsayad replied:
those requests can happen too late

fdkong commented on framework/src/variables/MooseVariableDataBase.C:
```diff
@@ -0,0 +1,827 @@
+
+    case Moose::Older:
+    {
+      needSolutionState(2);
```
Use enum would be better

lindsayad replied:
I like the integral argument for as many methods as possible. At some point we may want a solution state 3, and then adding an Olderer enum option would feel a little silly. I can remove these lines though

fdkong commented on framework/src/variables/MooseVariableDataBase.C:
```diff
@@ -0,0 +1,827 @@
+MooseVariableDataBase<OutputType>::setNodalValue(const OutputType & value, unsigned int idx)
+{
+  auto & dof_values = _vector_tags_dof_u[_solution_tag];
+  dof_values[idx] = value; // update variable nodal value
```
Should we put assert here? `idx` should be smaller than `dof_values.size()`

fdkong reviewed (changes requested):
Looks good to me. 

It is nice to merge FVData and FEData to remove code duplication 

Not sure whether or not we could improve the mix use of solution state and tag? For example, just stay with tag, and remove unnecessary state APIs if possible

lindsayad commented on framework/include/variables/MooseVariableDataBase.h:
```diff
@@ -0,0 +1,305 @@
+  const DoFValue & dofValues() const;
+  const DoFValue & dofValuesOld() const;
+  const DoFValue & dofValuesOlder() const;
+  const DoFValue & dofValuesPreviousNL() const;
```
maybe so. But I have already done much more in this PR than I was planning to haha

lindsayad commented on framework/include/variables/MooseVariableDataBase.h:
```diff
@@ -0,0 +1,305 @@
+  const DoFValue & nodalVectorTagValue(TagID tag) const;
+  const DoFValue & nodalMatrixTagValue(TagID tag) const;
+  const DoFValue & vectorTagDofValue(TagID tag) const;
+  const DoFValue & vectorTagDofValue(Moose::SolutionState state) const;
```
> Oh, we have one here. Can vectorTagDofValue replace dofValues and dofValuesXXX?

I was trying to accomplish this without modifying user code. If I remove these APIs I'll have to go update a lot of user code

lindsayad commented on framework/include/variables/MooseVariableDataBase.h:
```diff
@@ -0,0 +1,305 @@
+   */
+  void fetchDoFValues();
+  void zeroSizeDofValues();
+  void getArrayDoFValues(const NumericVector<Number> & sol,
```
This is migrated from `MooseVariableData`. I do not know what the original motivation for this was. Maybe @YaqiWang created this function

lindsayad commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -450,8 +450,75 @@ Coupleable::coupledValue(const std::string & var_name, unsigned int comp) const
+    var_sys.needSolutionState(1);
+  else if (tag_name == Moose::OLDER_SOLUTION_TAG)
+    var_sys.needSolutionState(2);
+}
```
You mean a `MooseEnum`? Because we are comparing against `TagNames`, e.g. `std::string`. These are not directly convertible to integers unless maybe we used a `MooseEnum`

lindsayad replied:
Or are you suggesting that we should change the `needSolutionState` signature to take an `enum` instead of an integer?

lindsayad replied:
either way we would still need a conversion from string to integral type

lindsayad committed:
Address Fande review

lindsayad committed:
Only evaluate mortar nodal aux kernel if var is nodal defined

lindsayad committed:
TagVectorArray -> TagVectorArrayVariable

fdkong reviewed (approved):
Looks good to me now

fdkong commented:
@YaqiWang  Do you have more comments?

lindsayad commented:
@fdkong can you resolve the conversations you started (make sure they're resolved to your liking)?

yjung-anl commented:
@lindsayad 

It is closed issue, but I have one question related with this issue. 
Recently, I added a new function named  `coupledVectorTagArrayValue` to use vector tag system for array variable in Griffin by myself before this issue was created and the associated PR merged. While creating the PR for my changes, I noticed this issue. 

Perhaps,  is `coupledVectorTagArrayValue` supported in the up-to-date version of framework? It seems that it only supports 'coupledVectorTagArrayDofValue'.

lindsayad commented:
correct, there is still no `coupledVectorTagArrayValue`. I had it at one point in this PR but then removed it because `coupledVectorTagArrayDofValue` was all that was needed for what I was trying to do. Related conversation starts [here](https://github.com/idaholab/moose/pull/20841#discussion_r858146332)
